narwhals/_spark/dataframe.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import Iterable
+from typing import Sequence
+
+from narwhals._spark.utils import parse_exprs_and_named_exprs
+from narwhals._spark.utils import translate_sql_api_dtype
+from narwhals.utils import Implementation
+from narwhals.utils import flatten
+from narwhals.utils import parse_columns_to_drop
+from narwhals.utils import parse_version
+
+if TYPE_CHECKING:
+    from pyspark.sql import DataFrame
+    from typing_extensions import Self
+
+    from narwhals._spark.expr import SparkExpr
+    from narwhals._spark.group_by import SparkLazyGroupBy
+    from narwhals._spark.namespace import SparkNamespace
+    from narwhals._spark.typing import IntoSparkExpr
+    from narwhals.dtypes import DType
+    from narwhals.typing import DTypes
+
+
+class SparkLazyFrame:
+    def __init__(
+        self,
+        native_dataframe: DataFrame,
+        *,
+        backend_version: tuple[int, ...],
+        dtypes: DTypes,
+    ) -> None:
+        self._native_frame = native_dataframe
+        self._backend_version = backend_version
+        self._implementation = Implementation.PYSPARK
+        self._dtypes = dtypes
+
+    def __native_namespace__(self) -> Any:  # pragma: no cover
+        if self._implementation is Implementation.PYSPARK:
+            return self._implementation.to_native_namespace()
+
+        msg = f"Expected pyspark, got: {type(self._implementation)}"  # pragma: no cover
+        raise AssertionError(msg)
+
+    def __narwhals_namespace__(self) -> SparkNamespace:
+        from narwhals._spark.namespace import SparkNamespace
+
+        return SparkNamespace(backend_version=self._backend_version, dtypes=self._dtypes)
+
+    def __narwhals_lazyframe__(self) -> Self:
+        return self
+
+    def _from_native_frame(self, df: DataFrame) -> Self:
+        return self.__class__(
+            df, backend_version=self._backend_version, dtypes=self._dtypes
+        )
+
+    @property
+    def columns(self) -> list[str]:
+        return self._native_frame.columns  # type: ignore[no-any-return]
+
+    def collect(self) -> Any:
+        import pandas as pd  # ignore-banned-import()
+
+        from narwhals._pandas_like.dataframe import PandasLikeDataFrame
+
+        return PandasLikeDataFrame(
+            native_dataframe=self._native_frame.toPandas(),
+            implementation=Implementation.PANDAS,
+            backend_version=parse_version(pd.__version__),
+            dtypes=self._dtypes,
+        )
```

Comment:
should we leave `collect` out for now?

Reply:
I think it is needed for collecting the results in the tests, right?  😕

Reply:
ah right!

ok for now, if we don't publicize pyspark support we can make changes later anyway if necessary

Reply:
Is there something in the current implementation that doesn't look ideal? (a part from this https://github.com/narwhals-dev/narwhals/pull/908#discussion_r1807362860)